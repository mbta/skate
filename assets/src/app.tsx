// We need to import the CSS so that webpack will load it.
// The MiniCssExtractPlugin is used to separate it out into
// its own CSS file.
declare function require(name: string): string
// tslint:disable-next-line
require("../css/app.scss")

// webpack automatically bundles all modules in your
// entry points. Those entry points can be configured
// in "webpack.config.js".
//
// Import dependencies
//
import "@babel/polyfill"
import "leaflet-defaulticon-compatibility/dist/leaflet-defaulticon-compatibility.webpack.css" // see https://github.com/Leaflet/Leaflet/issues/4968#issuecomment-483402699
import "leaflet/dist/leaflet.css"
import "phoenix_html"
import * as React from "react"
import ReactDOM from "react-dom"
import AppStateWrapper from "./components/appStateWrapper"

if (window.FS && window.username) {
  window.FS.identify(window.username, { displayName: window.username })
}

// Import local files
//
// Local files can be imported directly using relative paths, for example:
// import socket from "./socket"

ReactDOM.render(<AppStateWrapper />, document.getElementById("app"))
