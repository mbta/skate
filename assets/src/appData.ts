const appData = (): DOMStringMap | undefined => {
  const dataEl = document.getElementById("app")

  if (!dataEl) {
    return undefined
  }

  return dataEl.dataset
}

export default appData
