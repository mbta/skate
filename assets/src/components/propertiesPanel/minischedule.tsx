import React, { ReactElement, useContext } from "react"
import { StateDispatchContext } from "../../contexts/stateDispatchContext"
import { className } from "../../helpers/dom"
import {
  busFrontIcon,
  busRearIcon,
  filledCircleIcon,
  minusIcon,
  plusIcon,
  questionMarkIcon,
  triangleDownIcon,
  triangleUpIcon,
} from "../../helpers/icon"
import {
  useMinischeduleBlock,
  useMinischeduleRun,
} from "../../hooks/useMinischedule"
import { AsDirected, Block, Break, Piece, Run, Trip } from "../../minischedule"
import {
  directionOnLadder,
  getLadderDirectionForRoute,
  LadderDirections,
  VehicleDirection,
} from "../../models/ladderDirection"
import { DirectionId, RouteId, TripId } from "../../schedule"
import { formattedDuration, formattedScheduledTime } from "../../util/dateTime"
import Loading from "../loading"

export interface Props {
  activeTripId: TripId
}
export const MinischeduleRun = ({ activeTripId }: Props): ReactElement => {
  const run: Run | null | undefined = useMinischeduleRun(activeTripId)
  if (run === undefined) {
    return <Loading />
  } else if (run === null) {
    return <>No run found</>
  } else {
    return (
      <>
        <Header label="Run" value={run.id} />
        {run.activities.map((activity) =>
          isPiece(activity) ? (
            <Piece piece={activity} view="run" key={activity.start.time} />
          ) : (
            <Break break={activity} key={activity.startTime} />
          )
        )}
      </>
    )
  }
}

export const MinischeduleBlock = ({ activeTripId }: Props): ReactElement => {
  const block: Block | null | undefined = useMinischeduleBlock(activeTripId)
  if (block === undefined) {
    return <Loading />
  } else if (block === null) {
    return <>No block found</>
  } else {
    return (
      <>
        <Header label="Block" value={block.id} />
        {block.pieces.map((piece) => (
          <Piece piece={piece} view="block" key={piece.start.time} />
        ))}
      </>
    )
  }
}

const Header = ({ label, value }: { label: string; value: string }) => (
  <div className="m-minischedule__header">
    <span className="m-minischedule__header-label">{label}</span>
    {value}
  </div>
)

const Break = ({ break: breakk }: { break: Break }) => {
  const formattedBreakTime = formattedDuration(
    breakk.endTime - breakk.startTime
  )
  return (
    <Row text={`Break (${breakk.breakType})`} rightText={formattedBreakTime} />
  )
}

const Layover = ({
  currentTrip,
  nextTrip,
}: {
  currentTrip: Trip | AsDirected
  nextTrip?: Trip | AsDirected
}) => {
  if (!nextTrip) {
    return null
  }
  const layoverDuration = nextTrip.startTime - currentTrip.endTime
  if (layoverDuration === 0) {
    return null
  }

  return (
    <Row
      text="Layover"
      rightText={formattedDuration(layoverDuration)}
      extraClasses="m-minischedule__layover-row"
    />
  )
}

const Piece = ({ piece, view }: { piece: Piece; view: "run" | "block" }) => (
  <>
    {view === "block" ? (
      <div className="m-minischedule__run-header">{piece.runId}</div>
    ) : null}
    <div className="m-minischedule__piece-rows">
      <Row
        key="sign-on"
        icon={plusIcon()}
        text={
          piece.trips.length === 0 ||
          isAsDirected(piece.trips[0]) ||
          isDeadhead(piece.trips[0])
            ? "Start Time"
            : "Swing On"
        }
        rightText={formattedScheduledTime(piece.start.time)}
      />
      {piece.trips.map((trip, index) => {
        const sequence: "first" | "middle" | "last" =
          index === 0
            ? "first"
            : index === piece.trips.length - 1
            ? "last"
            : "middle"
        return (
          <React.Fragment key={trip.startTime}>
            {isTrip(trip) ? (
              <Trip trip={trip} sequence={sequence} />
            ) : (
              <AsDirected asDirected={trip} />
            )}
            {view === "run" ? (
              <Layover currentTrip={trip} nextTrip={piece.trips[index + 1]} />
            ) : null}
          </React.Fragment>
        )
      })}
      <Row
        key="sign-off"
        icon={minusIcon()}
        text={
          piece.trips.length === 0 ||
          isAsDirected(piece.trips[piece.trips.length - 1]) ||
          isDeadhead(piece.trips[piece.trips.length - 1])
            ? "Done"
            : "Swing Off"
        }
        rightText={formattedScheduledTime(piece.end.time)}
      />
    </div>
  </>
)

const Trip = ({
  trip,
  sequence,
}: {
  trip: Trip
  sequence: "first" | "middle" | "last"
}) => {
  if (isDeadhead(trip)) {
    return <DeadheadTrip trip={trip} sequence={sequence} />
  } else {
    return <RevenueTrip trip={trip} />
  }
}

const DeadheadTrip = ({
  trip,
  sequence,
}: {
  trip: Trip
  sequence: "first" | "middle" | "last"
}) => {
  const startTime: string = formattedScheduledTime(trip.startTime)
  if (sequence === "first") {
    return <Row icon={busFrontIcon()} text={"Pull Out"} rightText={startTime} />
  } else if (sequence === "last") {
    return <Row icon={busRearIcon()} text={"Pull Back"} rightText={startTime} />
  } else {
    return (
      <Row icon={filledCircleIcon()} text={"Deadhead"} rightText={startTime} />
    )
  }
}

const iconForDirectionOnLadder: (
  directionId: DirectionId | null,
  ladderDirections: LadderDirections,
  routeId: RouteId
) => ReactElement = (directionId, ladderDirections, routeId) => {
  if (directionId === null) {
    return questionMarkIcon()
  }

  const ladderDirection = getLadderDirectionForRoute(ladderDirections, routeId)
  if (
    directionOnLadder(directionId, ladderDirection) === VehicleDirection.Down
  ) {
    return triangleDownIcon()
  }
  return triangleUpIcon()
}

const RevenueTrip = ({ trip }: { trip: Trip }) => {
  const startTime: string = formattedScheduledTime(trip.startTime)
  const formattedVariant: string =
    trip.viaVariant !== null && trip.viaVariant !== "_" ? trip.viaVariant : ""
  const formattedRouteAndVariant: string = `${trip.routeId}_${formattedVariant}`
  const [{ ladderDirections }] = useContext(StateDispatchContext)

  const directionIcon =
    // Safe to assume routeId is not null, since if it were, we'd be
    // rendering a deadhead row instead.
    iconForDirectionOnLadder(trip.directionId, ladderDirections, trip.routeId!)

  return (
    <Row
      icon={directionIcon}
      text={
        <>
          {formattedRouteAndVariant}{" "}
          <span className="m-minischedule__headsign">{trip.headsign}</span>
        </>
      }
      rightText={startTime}
    />
  )
}

const AsDirected = ({ asDirected }: { asDirected: AsDirected }) => (
  <Row
    icon={busFrontIcon()}
    text={asDirected.kind === "rad" ? "Run as directed" : "Work as directed"}
    rightText={formattedScheduledTime(asDirected.startTime)}
  />
)

const Row = ({
  icon,
  text,
  rightText,
  extraClasses,
}: {
  icon?: ReactElement
  text: string | ReactElement
  rightText?: string
  extraClasses?: string
}) => (
  <div className={className(["m-minischedule__row", extraClasses])}>
    <div className="m-minischedule__icon">{icon}</div>
    <div className="m-minischedule__left-text">{text}</div>
    {rightText && <div className="m-minischedule__right-text">{rightText}</div>}
  </div>
)

const isPiece = (activity: Piece | Break): activity is Piece =>
  activity.hasOwnProperty("trips")

const isTrip = (trip: Trip | AsDirected): trip is Trip =>
  trip.hasOwnProperty("id")

const isAsDirected = (trip: Trip | AsDirected): trip is AsDirected =>
  !isTrip(trip)

const isDeadhead = (trip: Trip | AsDirected): boolean =>
  isTrip(trip) && trip.routeId == null
