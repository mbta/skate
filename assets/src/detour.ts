import { LatLngLiteral } from "leaflet"
import { Shape, ShapePoint } from "./schedule"

export interface DetourShape {
  coordinates: ShapePoint[]
  directions: {
    instruction: string
  }[]
}

export interface OriginalRoute {
  routeName: string
  routeDescription: string
  routeOrigin: string
  routeDirection: string
  shape: Shape
  center: LatLngLiteral
  zoom: number
}
