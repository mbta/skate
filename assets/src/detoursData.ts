import { array, Infer, number, string, type } from "superstruct"
import { DraftDetours } from "./models/detours"

export const DraftDetoursData = array(
  type({
    route: string(),
    direction: string(),
    name: string(),
    intersection: string(),
    active_since: number(),
  })
)
export type DraftDetoursData = Infer<typeof DraftDetoursData>

export const draftDetoursFromData = (
  draftDetoursData: DraftDetoursData
): DraftDetours =>
  draftDetoursData.map((draft) => ({
    route: draft.route,
    direction: draft.direction,
    name: draft.name,
    intersection: draft.intersection,
    activeSince: draft.active_since,
  }))
