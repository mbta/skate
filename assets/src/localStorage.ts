export const saveState = (stateKey: string, state: object) => {
  const serializedState = JSON.stringify(state)
  localStorage.setItem(stateKey, serializedState)
}

export const loadState = (stateKey: string): object | undefined => {
  const serializedState = localStorage.getItem(stateKey)

  if (serializedState === null) {
    return undefined
  }

  return JSON.parse(serializedState)
}
