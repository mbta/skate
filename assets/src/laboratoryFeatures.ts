import appData from "./appData"

const featureIsEnabled = (key: string): boolean => {
  const data = appData()

  if (data === undefined || data.laboratoryFeatures === undefined) {
    return false
  }

  const features = JSON.parse(data.laboratoryFeatures)
  if (features[key] === undefined) {
    return false
  }

  return features[key]
}

export default featureIsEnabled
