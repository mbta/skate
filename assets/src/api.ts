import "whatwg-fetch"
import appData from "./appData"
import { Block, Run } from "./minischedule"
import { reload } from "./models/browser"
import { blockFromData, runFromData } from "./models/minischeduleData"
import { swingsFromData } from "./models/swingsData"
import { NotificationId, NotificationState } from "./realtime.d"
import { RouteSettings } from "./routeSettings"
import {
  DirectionName,
  Route,
  RouteId,
  Shape,
  Swing,
  Timepoint,
  TripId,
} from "./schedule.d"

export interface RouteData {
  id: string
  direction_names: {
    "0": DirectionName
    "1": DirectionName
  }
  name: string
}

const checkResponseStatus = (response: Response) => {
  if (response.status === 200) {
    return response
  }

  if (Math.floor(response.status / 100) === 3 || response.status === 403) {
    // If the API sends us a redirect or forbidden, the user needs to
    // re-authenticate. Reload to go through the auth flow again.
    reload(true)
  }

  throw new Error(`Response error: ${response.status}`)
}

const parseJson = (response: Response) => response.json()

export const apiCall = <T>({
  url,
  parser,
  defaultResult,
}: {
  url: string
  parser: (data: any) => T
  defaultResult?: T
}): Promise<T> =>
  fetch(url)
    .then(checkResponseStatus)
    .then(parseJson)
    .then(({ data: data }: { data: any }) => parser(data))
    .catch((error) => {
      if (defaultResult === undefined) {
        throw error
      } else {
        return defaultResult
      }
    })

export const parseRouteData = ({
  id,
  direction_names,
  name,
}: RouteData): Route => ({
  id,
  directionNames: direction_names,
  name,
})

const parseRoutesData = (routesData: RouteData[]): Route[] =>
  routesData.map(parseRouteData)

export const fetchRoutes = (): Promise<Route[]> =>
  apiCall({
    url: "/api/routes",
    parser: parseRoutesData,
  })

export const fetchShapeForRoute = (routeId: RouteId): Promise<Shape[]> =>
  apiCall({
    url: `/api/shapes/route/${routeId}`,
    parser: (shapes: Shape[]) => shapes,
    defaultResult: [],
  })

export const fetchShapeForTrip = (tripId: TripId): Promise<Shape | null> =>
  apiCall({
    url: `/api/shapes/trip/${tripId}`,
    parser: (shape: Shape | null) => shape,
    defaultResult: null,
  })

export const fetchShuttleRoutes = (): Promise<Route[]> =>
  apiCall({
    url: "/api/shuttles",
    parser: parseRoutesData,
    defaultResult: [],
  })

export const fetchTimepointsForRoute = (
  routeId: RouteId
): Promise<Timepoint[]> =>
  apiCall({
    url: `/api/routes/${routeId}`,
    parser: (timepoints: Timepoint[]) => timepoints,
    defaultResult: [],
  })

export const fetchMinischeduleRun = (tripId: TripId): Promise<Run | null> =>
  apiCall({
    url: `/api/minischedule/run/${tripId}`,
    parser: nullableParser(runFromData),
    defaultResult: null,
  })

export const fetchMinischeduleBlock = (tripId: TripId): Promise<Block | null> =>
  apiCall({
    url: `/api/minischedule/block/${tripId}`,
    parser: nullableParser(blockFromData),
    defaultResult: null,
  })

export const fetchNearestIntersection = (
  latitude: number,
  longitude: number
): Promise<string | null> =>
  apiCall({
    url: `/api/intersection?latitude=${latitude}&longitude=${longitude}`,
    parser: nullableParser((intersection: string) => intersection),
    defaultResult: null,
  })

export const fetchSwings = (): Promise<Swing[] | null> =>
  apiCall({
    url: "/api/swings",
    parser: nullableParser(swingsFromData),
    defaultResult: null,
  })

export const putNotificationReadState = (
  newReadState: NotificationState,
  notificationIds: NotificationId[]
): void => {
  const url = `/api/notification_read_state?new_state=${newReadState}&notification_ids=${notificationIds.join(
    ","
  )}`
  fetch(url, {
    method: "PUT",
    headers: {
      "Content-Type": "application/json",
      "x-csrf-token": getCsrfToken(),
    },
  })
}

export const putUserSetting = (field: string, value: string): void => {
  const url = `/api/user_settings?field=${field}&value=${value}`
  fetch(url, {
    method: "PUT",
    headers: {
      "Content-Type": "application/json",
      "x-csrf-token": getCsrfToken(),
    },
  })
}

export const putRouteSettings = (routeSettings: RouteSettings): void => {
  fetch("/api/route_settings", {
    method: "PUT",
    headers: {
      "Content-Type": "application/json",
      "x-csrf-token": getCsrfToken(),
    },
    body: JSON.stringify(routeSettings),
  })
}

const getCsrfToken = (): string => appData()?.csrfToken || ""

export const nullableParser =
  <Data, T>(parser: (data: Data) => T): ((data: Data | null) => T | null) =>
  (data: Data | null) =>
    data === null ? null : parser(data)
