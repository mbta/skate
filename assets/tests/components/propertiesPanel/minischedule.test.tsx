import React from "react"
import renderer from "react-test-renderer"
import {
  MinischeduleBlock,
  MinischeduleRun,
} from "../../../src/components/propertiesPanel/minischedule"
import {
  useMinischeduleBlock,
  useMinischeduleRun,
} from "../../../src/hooks/useMinischedule"
import { Break, Piece, Trip } from "../../../src/minischedule"

jest.mock("../../../src/hooks/useMinischedule", () => ({
  __esModule: true,
  useMinischeduleRun: jest.fn(),
  useMinischeduleBlock: jest.fn(),
}))

const breakk: Break = {
  breakType: "Break",
  startTime: 10,
  endTime: 11,
}

const nonrevenueTrip: Trip = {
  id: "nonrevenue",
  blockId: "block",
  routeId: null,
  headsign: null,
  directionId: null,
  viaVariant: null,
  runId: null,
  startTime: 0,
  endTime: 1,
}

const revenueTrip: Trip = {
  id: "trip",
  blockId: "block",
  routeId: "R",
  headsign: "Revenue",
  directionId: 1,
  viaVariant: "X",
  runId: "run",
  startTime: 0,
  endTime: 1,
}

const piece: Piece = {
  runId: "run",
  blockId: "block",
  start: {
    time: 20,
    place: "start",
    midRoute: false,
  },
  trips: [revenueTrip],
  end: {
    time: 21,
    place: "end",
    midRoute: false,
  },
}

describe("MinischeduleRun", () => {
  test("renders the loading state", () => {
    ;(useMinischeduleRun as jest.Mock).mockImplementationOnce(() => undefined)
    const tree = renderer
      .create(<MinischeduleRun activeTripId={"trip"} />)
      .toJSON()

    expect(tree).toMatchSnapshot()
  })

  test("renders a not found state", () => {
    ;(useMinischeduleRun as jest.Mock).mockImplementationOnce(() => null)
    const tree = renderer
      .create(<MinischeduleRun activeTripId={"trip"} />)
      .toJSON()

    expect(tree).toMatchSnapshot()
  })

  test("renders a run", () => {
    ;(useMinischeduleRun as jest.Mock).mockImplementationOnce(() => ({
      id: "run",
      activities: [breakk, piece],
    }))
    const tree = renderer
      .create(<MinischeduleRun activeTripId={"trip"} />)
      .toJSON()

    expect(tree).toMatchSnapshot()
  })
})

describe("MinischeduleBlock", () => {
  test("renders the loading state", () => {
    ;(useMinischeduleBlock as jest.Mock).mockImplementationOnce(() => undefined)
    const tree = renderer
      .create(<MinischeduleBlock activeTripId={"trip"} />)
      .toJSON()

    expect(tree).toMatchSnapshot()
  })

  test("renders a not found state", () => {
    ;(useMinischeduleBlock as jest.Mock).mockImplementationOnce(() => null)
    const tree = renderer
      .create(<MinischeduleBlock activeTripId={"trip"} />)
      .toJSON()

    expect(tree).toMatchSnapshot()
  })

  test("renders a block", () => {
    ;(useMinischeduleBlock as jest.Mock).mockImplementationOnce(() => ({
      id: "block",
      pieces: [piece],
    }))
    const tree = renderer
      .create(<MinischeduleBlock activeTripId={"trip"} />)
      .toJSON()

    expect(tree).toMatchSnapshot()
  })

  test("renders pulls and deadheads", () => {
    const deadheadPiece: Piece = {
      ...piece,
      trips: [
        { ...nonrevenueTrip, id: "pullout" },
        { ...nonrevenueTrip, id: "deadhead" },
        { ...nonrevenueTrip, id: "pullback" },
      ],
    }
    ;(useMinischeduleBlock as jest.Mock).mockImplementationOnce(() => ({
      id: "block",
      pieces: [deadheadPiece],
    }))
    const tree = renderer
      .create(<MinischeduleBlock activeTripId={"trip"} />)
      .toJSON()

    expect(tree).toMatchSnapshot()
  })
})
