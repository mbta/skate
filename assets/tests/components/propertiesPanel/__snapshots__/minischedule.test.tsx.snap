// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MinischeduleBlock renders a block 1`] = `
Array [
  <div
    className="m-minischedule__header"
  >
    <span
      className="m-minischedule__header-label"
    >
      Block
    </span>
    block
  </div>,
  <div
    className="m-minischedule__run-header"
  >
    run
  </div>,
  <div
    className="m-minischedule__piece-rows"
  >
    <div
      className="m-minischedule__row"
    >
      <div
        className="m-minischedule__icon"
      >
        <span
          className=""
          dangerouslySetInnerHTML={
            Object {
              "__html": "SVG",
            }
          }
        />
      </div>
      <div
        className="m-minischedule__left-text"
      >
        Swing On
      </div>
      <div
        className="m-minischedule__right-text"
      >
        12:30am
      </div>
    </div>
    <div
      className="m-minischedule__row"
    >
      <div
        className="m-minischedule__icon"
      >
        <span
          className=""
          dangerouslySetInnerHTML={
            Object {
              "__html": "SVG",
            }
          }
        />
      </div>
      <div
        className="m-minischedule__left-text"
      >
        R_X
         
        <span
          className="m-minischedule__headsign"
        >
          Revenue
        </span>
      </div>
      <div
        className="m-minischedule__right-text"
      >
        12:00am
      </div>
    </div>
    <div
      className="m-minischedule__row"
    >
      <div
        className="m-minischedule__icon"
      >
        <span
          className=""
          dangerouslySetInnerHTML={
            Object {
              "__html": "SVG",
            }
          }
        />
      </div>
      <div
        className="m-minischedule__left-text"
      >
        Swing Off
      </div>
      <div
        className="m-minischedule__right-text"
      >
        12:30am
      </div>
    </div>
  </div>,
]
`;

exports[`MinischeduleBlock renders a not found state 1`] = `"No block found"`;

exports[`MinischeduleBlock renders pulls and deadheads 1`] = `
Array [
  <div
    className="m-minischedule__header"
  >
    <span
      className="m-minischedule__header-label"
    >
      Block
    </span>
    block
  </div>,
  <div
    className="m-minischedule__run-header"
  >
    run
  </div>,
  <div
    className="m-minischedule__piece-rows"
  >
    <div
      className="m-minischedule__row"
    >
      <div
        className="m-minischedule__icon"
      >
        <span
          className=""
          dangerouslySetInnerHTML={
            Object {
              "__html": "SVG",
            }
          }
        />
      </div>
      <div
        className="m-minischedule__left-text"
      >
        Start Time
      </div>
      <div
        className="m-minischedule__right-text"
      >
        12:30am
      </div>
    </div>
    <div
      className="m-minischedule__row"
    >
      <div
        className="m-minischedule__icon"
      >
        <span
          className=""
          dangerouslySetInnerHTML={
            Object {
              "__html": "SVG",
            }
          }
        />
      </div>
      <div
        className="m-minischedule__left-text"
      >
        Pull Out
      </div>
      <div
        className="m-minischedule__right-text"
      >
        12:00am
      </div>
    </div>
    <div
      className="m-minischedule__row"
    >
      <div
        className="m-minischedule__icon"
      >
        <span
          className=""
          dangerouslySetInnerHTML={
            Object {
              "__html": "SVG",
            }
          }
        />
      </div>
      <div
        className="m-minischedule__left-text"
      >
        Deadhead
      </div>
      <div
        className="m-minischedule__right-text"
      >
        12:03am
      </div>
    </div>
    <div
      className="m-minischedule__row"
    >
      <div
        className="m-minischedule__icon"
      >
        <span
          className=""
          dangerouslySetInnerHTML={
            Object {
              "__html": "SVG",
            }
          }
        />
      </div>
      <div
        className="m-minischedule__left-text"
      >
        Pull Back
      </div>
      <div
        className="m-minischedule__right-text"
      >
        12:14am
      </div>
    </div>
    <div
      className="m-minischedule__row"
    >
      <div
        className="m-minischedule__icon"
      >
        <span
          className=""
          dangerouslySetInnerHTML={
            Object {
              "__html": "SVG",
            }
          }
        />
      </div>
      <div
        className="m-minischedule__left-text"
      >
        Done
      </div>
      <div
        className="m-minischedule__right-text"
      >
        12:30am
      </div>
    </div>
  </div>,
]
`;

exports[`MinischeduleBlock renders the loading state 1`] = `"loading..."`;

exports[`MinischeduleBlock renders trips in both directions, or missing direction 1`] = `
Array [
  <div
    className="m-minischedule__header"
  >
    <span
      className="m-minischedule__header-label"
    >
      Block
    </span>
    block
  </div>,
  <div
    className="m-minischedule__run-header"
  >
    run
  </div>,
  <div
    className="m-minischedule__piece-rows"
  >
    <div
      className="m-minischedule__row"
    >
      <div
        className="m-minischedule__icon"
      >
        <span
          className=""
          dangerouslySetInnerHTML={
            Object {
              "__html": "SVG",
            }
          }
        />
      </div>
      <div
        className="m-minischedule__left-text"
      >
        Swing On
      </div>
      <div
        className="m-minischedule__right-text"
      >
        12:30am
      </div>
    </div>
    <div
      className="m-minischedule__row"
    >
      <div
        className="m-minischedule__icon"
      >
        <span
          className=""
          dangerouslySetInnerHTML={
            Object {
              "__html": "SVG",
            }
          }
        />
      </div>
      <div
        className="m-minischedule__left-text"
      >
        R_X
         
        <span
          className="m-minischedule__headsign"
        >
          Revenue
        </span>
      </div>
      <div
        className="m-minischedule__right-text"
      >
        12:00am
      </div>
    </div>
    <div
      className="m-minischedule__row"
    >
      <div
        className="m-minischedule__icon"
      >
        <span
          className=""
          dangerouslySetInnerHTML={
            Object {
              "__html": "SVG",
            }
          }
        />
      </div>
      <div
        className="m-minischedule__left-text"
      >
        R_X
         
        <span
          className="m-minischedule__headsign"
        >
          Revenue The Other Direction
        </span>
      </div>
      <div
        className="m-minischedule__right-text"
      >
        12:05am
      </div>
    </div>
    <div
      className="m-minischedule__row"
    >
      <div
        className="m-minischedule__icon"
      >
        <span
          className=""
          dangerouslySetInnerHTML={
            Object {
              "__html": "SVG",
            }
          }
        />
      </div>
      <div
        className="m-minischedule__left-text"
      >
        R_X
         
        <span
          className="m-minischedule__headsign"
        >
          Revenue Missing Direction
        </span>
      </div>
      <div
        className="m-minischedule__right-text"
      >
        12:10am
      </div>
    </div>
    <div
      className="m-minischedule__row"
    >
      <div
        className="m-minischedule__icon"
      >
        <span
          className=""
          dangerouslySetInnerHTML={
            Object {
              "__html": "SVG",
            }
          }
        />
      </div>
      <div
        className="m-minischedule__left-text"
      >
        Swing Off
      </div>
      <div
        className="m-minischedule__right-text"
      >
        12:30am
      </div>
    </div>
  </div>,
]
`;

exports[`MinischeduleRun renders a not found state 1`] = `"No run found"`;

exports[`MinischeduleRun renders a run 1`] = `
Array [
  <div
    className="m-minischedule__header"
  >
    <span
      className="m-minischedule__header-label"
    >
      Run
    </span>
    run
  </div>,
  <div
    className="m-minischedule__row"
  >
    <div
      className="m-minischedule__icon"
    />
    <div
      className="m-minischedule__left-text"
    >
      Break (Paid meal before)
    </div>
    <div
      className="m-minischedule__right-text"
    >
      30m
    </div>
  </div>,
  <div
    className="m-minischedule__piece-rows"
  >
    <div
      className="m-minischedule__row"
    >
      <div
        className="m-minischedule__icon"
      >
        <span
          className=""
          dangerouslySetInnerHTML={
            Object {
              "__html": "SVG",
            }
          }
        />
      </div>
      <div
        className="m-minischedule__left-text"
      >
        Swing On
      </div>
      <div
        className="m-minischedule__right-text"
      >
        12:30am
      </div>
    </div>
    <div
      className="m-minischedule__row"
    >
      <div
        className="m-minischedule__icon"
      >
        <span
          className=""
          dangerouslySetInnerHTML={
            Object {
              "__html": "SVG",
            }
          }
        />
      </div>
      <div
        className="m-minischedule__left-text"
      >
        R_X
         
        <span
          className="m-minischedule__headsign"
        >
          Revenue
        </span>
      </div>
      <div
        className="m-minischedule__right-text"
      >
        12:00am
      </div>
    </div>
    <div
      className="m-minischedule__row m-minischedule__layover-row"
    >
      <div
        className="m-minischedule__icon"
      />
      <div
        className="m-minischedule__left-text"
      >
        Layover
      </div>
      <div
        className="m-minischedule__right-text"
      >
        5m
      </div>
    </div>
    <div
      className="m-minischedule__row"
    >
      <div
        className="m-minischedule__icon"
      >
        <span
          className=""
          dangerouslySetInnerHTML={
            Object {
              "__html": "SVG",
            }
          }
        />
      </div>
      <div
        className="m-minischedule__left-text"
      >
        R_X
         
        <span
          className="m-minischedule__headsign"
        >
          Revenue The Other Direction
        </span>
      </div>
      <div
        className="m-minischedule__right-text"
      >
        12:05am
      </div>
    </div>
    <div
      className="m-minischedule__row"
    >
      <div
        className="m-minischedule__icon"
      >
        <span
          className=""
          dangerouslySetInnerHTML={
            Object {
              "__html": "SVG",
            }
          }
        />
      </div>
      <div
        className="m-minischedule__left-text"
      >
        Swing Off
      </div>
      <div
        className="m-minischedule__right-text"
      >
        12:30am
      </div>
    </div>
  </div>,
]
`;

exports[`MinischeduleRun renders a run with no layovers between trips 1`] = `
Array [
  <div
    className="m-minischedule__header"
  >
    <span
      className="m-minischedule__header-label"
    >
      Run
    </span>
    run
  </div>,
  <div
    className="m-minischedule__piece-rows"
  >
    <div
      className="m-minischedule__row"
    >
      <div
        className="m-minischedule__icon"
      >
        <span
          className=""
          dangerouslySetInnerHTML={
            Object {
              "__html": "SVG",
            }
          }
        />
      </div>
      <div
        className="m-minischedule__left-text"
      >
        Swing On
      </div>
      <div
        className="m-minischedule__right-text"
      >
        12:30am
      </div>
    </div>
    <div
      className="m-minischedule__row"
    >
      <div
        className="m-minischedule__icon"
      >
        <span
          className=""
          dangerouslySetInnerHTML={
            Object {
              "__html": "SVG",
            }
          }
        />
      </div>
      <div
        className="m-minischedule__left-text"
      >
        R_X
         
        <span
          className="m-minischedule__headsign"
        >
          Revenue
        </span>
      </div>
      <div
        className="m-minischedule__right-text"
      >
        12:00am
      </div>
    </div>
    <div
      className="m-minischedule__row"
    >
      <div
        className="m-minischedule__icon"
      >
        <span
          className=""
          dangerouslySetInnerHTML={
            Object {
              "__html": "SVG",
            }
          }
        />
      </div>
      <div
        className="m-minischedule__left-text"
      >
        R_X
         
        <span
          className="m-minischedule__headsign"
        >
          Revenue The Other Direction
        </span>
      </div>
      <div
        className="m-minischedule__right-text"
      >
        12:00am
      </div>
    </div>
    <div
      className="m-minischedule__row"
    >
      <div
        className="m-minischedule__icon"
      >
        <span
          className=""
          dangerouslySetInnerHTML={
            Object {
              "__html": "SVG",
            }
          }
        />
      </div>
      <div
        className="m-minischedule__left-text"
      >
        Swing Off
      </div>
      <div
        className="m-minischedule__right-text"
      >
        12:30am
      </div>
    </div>
  </div>,
]
`;

exports[`MinischeduleRun renders the loading state 1`] = `"loading..."`;
